plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/declarations/FirKotlinUVariable.kt

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.uast.kotlin
+
+import com.intellij.psi.*
+import org.jetbrains.kotlin.asJava.elements.KtLightElement
+import org.jetbrains.kotlin.fir.declarations.FirProperty
+import org.jetbrains.kotlin.fir.declarations.FirValueParameter
+import org.jetbrains.kotlin.fir.declarations.FirVariable
+import org.jetbrains.kotlin.fir.expressions.FirExpression
+import org.jetbrains.kotlin.idea.fir.low.level.api.api.getOrBuildFirOfType
+import org.jetbrains.kotlin.idea.fir.low.level.api.api.getResolveState
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.uast.*
+import org.jetbrains.uast.internal.acceptList
+import org.jetbrains.uast.visitor.UastVisitor
+
+sealed class AbstractFirKotlinUVariable(
+    givenParent: UElement?
+) : FirKotlinAbstractUElement(givenParent), PsiVariable, UVariableEx, UAnchorOwner {
+    override val uAnnotations: List<UAnnotation>
+        get() {
+            // TODO: Not yet implemented
+            return emptyList()
+        }
+
+    override fun getNameIdentifier(): PsiIdentifier {
+        val kotlinOrigin = (sourcePsi as? KtLightElement<*, *>)?.kotlinOrigin
+        return FirUastLightIdentifier(psi, kotlinOrigin as? KtDeclaration)
+    }
+
+    override fun getContainingFile(): PsiFile = unwrapFakeFileForLightClass(psi.containingFile)
+
+    override val uastAnchor: UIdentifier? by lz {
+        val identifierSourcePsi = when (val sourcePsi = sourcePsi) {
+            is KtNamedDeclaration -> sourcePsi.nameIdentifier
+            is KtTypeReference -> sourcePsi.typeElement?.let {
+                // receiver param in extension function
+                (it as? KtUserType)?.referenceExpression?.getIdentifier() ?: it
+            } ?: sourcePsi
+            is KtNameReferenceExpression -> sourcePsi.getReferencedNameElement()
+            is KtBinaryExpression, is KtCallExpression -> null // e.g. `foo("Lorem ipsum") ?: foo("dolor sit amet")`
+            is KtDestructuringDeclaration -> sourcePsi.valOrVarKeyword
+            is KtLambdaExpression -> sourcePsi.functionLiteral.lBrace
+            else -> sourcePsi
+        } ?: return@lz null
+        FirKotlinUIdentifier(nameIdentifier, identifierSourcePsi, this)
+    }
+
+    override val typeReference: UTypeReferenceExpression? by lz {
+        (sourcePsi as? KtCallableDeclaration)?.typeReference?.let {
+            FirKotlinUTypeReferenceExpression(it, this) { type }
+        }
+    }
+
+    private val kotlinOrigin by lz { getKotlinMemberOrigin(psi.originalElement) ?: sourcePsi }
+
+    protected val firVariable: FirVariable<*>? by lz {
```

Comment:
This is similar to `FirFunction`. To convert variable initializing expression, this is the last point where we can get FIR element.

Reply:
Ditto. The use of FIR is not necessary at the moment.